x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -419,7 +420,13 @@ protected LogicalPlan rule(UnaryPlan plan) {
             // Agg with underlying Project (group by on sub-queries)
             if (plan instanceof Aggregate a) {
                 if (child instanceof Project p) {
-                    plan = new Aggregate(a.source(), p.child(), a.groupings(), combineProjections(a.aggregates(), p.projections()));
+                    plan = new Aggregate(
+                        a.source(),
+                        p.child(),
+                        // After substitutions groupings can only contain attributes.
+                        combineUpperGroupingsAndLowerProjections((List<? extends Attribute>) (List<?>) a.groupings(), p.projections()),
+                        combineProjections(a.aggregates(), p.projections())
+                    );
```

Comment:
If we really expect Attributes at this point, let's make this intention clear: I've changed the ugly double casting plus made the comment more clear; also, a proper exception is thrown if our expectation is not met (vs. letting a CCE being thrown in case the list of groupings is not a list of Attributes).

```suggestion
                    var groupings = a.groupings();
                    List<Attribute> groupingAttrs = new ArrayList<>(a.groupings().size());
                    for (Expression grouping : groupings) {
                        if (grouping instanceof Attribute attribute) {
                            groupingAttrs.add(attribute);
                        } else {
                            // After "Substitutions" batch, the groupings of an Aggregate can only be attributes
                            throw new EsqlIllegalArgumentException("Expected an Attribute, got {}", grouping);
                        }
                    }
                    plan = new Aggregate(
                        a.source(),
                        p.child(),
                        combineUpperGroupingsAndLowerProjections(groupingAttrs, p.projections()),
                        combineProjections(a.aggregates(), p.projections())
                    );
```

Reply:
Yeah, that's better, thank you!